docs: Improve Getting Started (#6551)
nextauthjs/next-auth: docs/docs/getting-started/02-oauth-tutorial.mdx

import startAppAndSignInImg from "./img/getting-started-app-start.png"
import githubAuthCredentials from "./img/getting-started-github-auth.png"
import nextAuthUserLoggedIn from "./img/getting-started-nextauth-success.png"
import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"

We understand that authentication is tedious - It is a rabbit hole that is super easy to get lost on. The goal of making Auth.js is to help you easily add authentication to your project with just a few lines of code.

> The easiest way is to set up Auth.js with an [OAuth](https://en.wikipedia.org/wiki/OAuth) provider. In this tutorial, we'll be setting Auth.js in an application with a framework of your choice to be able to log in with **GitHub**.

Reviewer: We should have our own OAuth page under Concepts, taking the graph from https://next-auth.js.org/configuration/providers/oauth

Author: We have that page https://authjs.dev/concepts/oauth Are you suggesting to link it here?

> The easiest way is to set up Auth.js with an [OAuth](https://en.wikipedia.org/wiki/OAuth) provider. In this tutorial, we'll be setting Auth.js in an application with a framework of your choice to be able to log in with **GitHub**.

Reviewer: ```suggestion
The easiest way is to set up Auth.js with an [OAuth](/concepts/oauth) provider. In this tutorial, we'll be setting Auth.js in an application with a framework of your choice to be able to log in with **GitHub**.
```

})
```

You may notice we are using environment variables in the code example above. We take the value of `GITHUB_ID` and `GITHUB_SECRET` from the GitHub Developer OAuth Portal. See [Configuring OAuth Provider](/getting-started/oauth-tutorial#2-configuring-oauth-provider) section on how to get those.

:::info

Behind the scenes, this creates all the relevant OAuth API routes within `/api/auth/*` so that auth API requests to:

- `/api/auth/callback`
- `/api/auth/signin`
- `/api/auth/signout`
- etc.

> can be handled by Auth.js. In this way, Auth.js stays in charge of the whole application's authentication request/response flow.

Reviewer: ```suggestion
can be handled by NextAuth.js. In this way, NextAuth.js stays in charge of the whole application's authentication request/response flow.
```

> Auth.js is fullly customizable - [our guides section](/guides/overview) will teach you how to set it up to handle auth in different ways. All the possible configuration options are [listed here](/reference/configuration/auth-config).

Reviewer: ```suggestion
NextAuth.js is fully customizable - [our guides section](/guides/overview) teaches you how to set it up to handle auth in different ways. All the possible configuration options are [listed here](/reference/configuration/auth-config).
```

})
```

:::info

Behind the scenes, this creates all the relevant OAuth API routes within `/api/auth/*` so that auth API requests to:

- [GET `/api/auth/signin`](https://authjs.dev/reference/rest-api#get--apiauthsignin)
- [POST `/api/auth/signin/:provider`](https://authjs.dev/reference/rest-api#post--apiauthsigninprovider)
- [GET/POST `/api/auth/callback/:provider`](https://authjs.dev/reference/rest-api#get--post--apiauthcallbackprovider)
- [GET `/api/auth/signout`](https://authjs.dev/reference/rest-api#get--apiauthsignout)
- [POST `/api/auth/signout`](https://authjs.dev/reference/rest-api#post--apiauthsignout)
- [GET `/api/auth/session`](https://authjs.dev/reference/rest-api#get--apiauthsession)
- [GET `/api/auth/csrf`](https://authjs.dev/reference/rest-api#get--apiauthcsrf)
- [GET `/api/auth/providers`](https://authjs.dev/reference/rest-api#get--apiauthproviders)

> can be handled by NextAuth.js. In this way, NextAuth.js stays in charge of the whole application's authentication request/response flow.

Reviewer 2: ```suggestion
can be handled by Auth.js. In this way, Auth.js stays in charge of the whole application's authentication request/response flow.
```

> NextAuth.js is fully customizable - [our guides section](/guides/overview) teaches you how to set it up to handle auth in different ways. All the possible configuration options are [listed here](/reference/configuration/auth-config).

Reviewer 2: ```suggestion
Auth.js is fully customizable - [our guides section](/guides/overview) teaches you how to set it up to handle auth in different ways. All the possible configuration options are [listed here](/reference/configuration/auth-config).
```

or https://generate-secret.vercel.app/32 to generate a random value for it.

### Exposing the session via `SessionProvider`:

> To use `useSession` first you'll need to expose the session context - [`<SessionProvider />`](/reference/react/#sessionprovider) - at the top level of your application:

Reviewer: `useSession` wasn't mentioned at this point, we should explain in a sentence what that is.

}
```

> Instances of `useSession` (more on it in the next section) will have access to the session data and status. The `<SessionProvider />` also keep the session updated and synced between browser tabs and windows. 💪🏽

Reviewer: Move this up to satisfy https://github.com/nextauthjs/next-auth/pull/6551/files#r1089949056

:::tip
Check our [client docs](/reference/react/) to learn all the available options for handling sessions on the browser.
:::

### Consuming the session via hooks

> Auth.js exposes a [`useSession()`](/reference/react/#usesession) React Hook so that you can check if someone is logged in:

Reviewer: ```suggestion
NextAuth.js exposes a [`useSession()`](/reference/react/#usesession) React Hook so that you can check if someone is logged in:
```

}
```

> You can use the `useSession` hook from anywhere in your application (E.g. in a header component). Behind the scenes, the hook will connect to the `<SessionProvider />` to read the current user session.

Reviewer: Could drop a link to https://reactjs.org/docs/context.html

### Protecting API Routes

> Protecting your API Routes (blocking unauthorized access to resources) is easy! You can use [`getServerSession()`](/reference/nextjs#getserversession) to know whether a session exists or not:

Reviewer: Let's rephrase to remove "easy". Nothing is if you are just starting out! :green_heart:

Author: Totally!

```ts title="pages/api/movies/list.ts"
import { getServerSession } from "next-auth/next"

> import authOptions from "@/pages/api/auth/[...nextauth]"

Reviewer: ```suggestion
import authOptions from "../auth/[...nextauth]"
```

Let's not assume anything about the user's setup. They can change it if they want to.

  return (
    <>
      <p>Not signed in.</p>
      <button onClick={() => signIn("github")}>Sign in</button>
    </>
  )
}
```

### Protecting API Routes

> To protecting your API Routes (blocking unauthorized access to resources), you can use [`getServerSession()`](/reference/nextjs#getserversession) to know whether a session exists or not:

Reviewer 2: ```suggestion
To protect your API Routes (blocking unauthorized access to resources), you can use [`getServerSession()`](/reference/nextjs#getserversession) to know whether a session exists or not:
```

}
```

  </TabItem>
  <TabItem value="sveltekit" label="SvelteKit">
TODO: SvelteKit
  </TabItem>
  <TabItem value="solidstart" label="SolidStart">
TODO: SolidStart
  </TabItem>
</Tabs>

## 2. Configuring OAuth Provider

> Ok, we have our app set up with Auth.js, however, if you run the app right now, it won't work as we haven't configured our OAuth provider (**GitHub**) yet.

Reviewer: ```suggestion
Ok, we have our app set up with NextAuth.js, however, if you run the app right now, it won't work as we haven't configured our OAuth provider (**GitHub**) yet.
```

:::info
When using OAuth you're asking for a third-party service (in this case GitHub, although it could be Google, Twitter, etc...) to handle user authentication for your app.
:::

> We need to register our new app in GitHub, so that when Auth.js forwards the authorization requests to it, GitHub can recognize your application and prompt the user to sign in.

Reviewer: ```suggestion
We need to register our new app in GitHub, so that when NextAuth.js forwards the authorization requests to it, GitHub can recognize your application and prompt the user to sign in.
```

<img src={creatingOauthAppImg} />

> Log in to **GitHub**, go to `Settings / Developers / OAuth Apps` and click "New OAuth App"

Reviewer: ```suggestion
Log in to **GitHub**, go to [`Settings / Developers / OAuth Apps`](https://github.com/settings/developers) and click "New OAuth App"
```

}
```

  </TabItem>
  <TabItem value="sveltekit" label="SvelteKit">
TODO: SvelteKit
  </TabItem>
  <TabItem value="solidstart" label="SolidStart">
TODO: SolidStart
  </TabItem>
  <TabItem value="core" label="Vanilla (No Framework)">
TODO Core
  </TabItem>

> </Tabs>

Reviewer 2: I guess this should be addressed before we merge right?

## 2. Configuring OAuth Provider

> Ok, we have our app set up with NextAuth.js, however, if you run the app right now, it won't work as we haven't configured our OAuth provider (**GitHub**) yet.

Reviewer 2: ```suggestion
Ok, we have our app set up with Auth.js; however, if you run the app right now, it won't work as we haven't configured our OAuth provider (**GitHub**) yet.
```

:::info
When using OAuth you're asking for a third-party service (in this case GitHub, although it could be Google, Twitter, etc...) to handle user authentication for your app.
:::

> We need to register our new app in GitHub, so that when NextAuth.js forwards the authorization requests to it, GitHub can recognize your application and prompt the user to sign in.

Reviewer 2: ```suggestion
We need to register our new app in GitHub, so that when Auth.js forwards the authorization requests to it, GitHub can recognize your application and prompt the user to sign in.
```

[origin]/api/auth/callback/[provider]
```

In this case, given we want to try our authentication working locally on our machine and we're using **GitHub** as our OAuth provider, it'll be:

<Tabs groupId="frameworks">
  <TabItem value="next" label="Next.js" default>

```
http://localhost:3000/api/auth/callback/github
```

> 

Reviewer: Let's add a note here where the last portion of the URL (`github`) comes from, and what is the relation with `signIn("github")`.

Not sure what's the best way to document this, but it's the provider configuration's `id` property.

:::info

> Auth.js will already magically create this API endpoint for you when we start the application later. Note that because we're using Next.js, locally it starts our server on port `3000`. Hence, the origin is `http://localhost:3000`.

Reviewer: ```suggestion
NextAuth.js will already create this API endpoint for you when we start the application later. Note that because we're using Next.js, locally it starts our server on port `3000` by default. Hence, the origin is `http://localhost:3000`.
```

:::

  </TabItem>
  <TabItem value="sveltekit" label="SvelteKit">

  ```
  http://localhost:5173/auth/callback/github
  ```

  </TabItem>
  <TabItem value="solidstart" label="SolidStart">
    TODO SolidStart
  </TabItem>
</Tabs>
To register, tap on "Register application" button.

The next screen shows all the configurations for your newly created OAuth app. For now, we need two things from it - the **Client ID** and **Client Secret**:

<img src={gettingClientIdSecretImg} />

The Client ID is always there, a public identifier of your OAuth application within GitHub. Click on the **Generate a new client Secret** button and should be presented with a new string (which is just a randomized string).

:::warning

> 🔥 Keep both your Client ID and Client Secret secure and never expose them to the public or shared them with people outside your organization. With them, a malicious actor could hijack your application and cause you and your user serious problems!

Reviewer: ```suggestion
Keep both your Client ID and Client Secret secure and never expose them to the public or share them with people outside your organization. With them, a malicious actor could hijack your application and cause you and your user serious problems!
```

:::

Cool! We have finished configuring our OAuth provider, now let's wire all together so we can finally see authentication working in our app!

:::info

> As noted previously, Auth.js has built-in support for multiple OAuth providers, <a href="">here is the full list</a>. You can also easily build your own in case the provider you need is not on the list.

Reviewer: ```suggestion
As noted previously, NextAuth.js has built-in support for multiple OAuth providers, <a href="">here is the full list</a>. You can also easily build your own in case the provider you need is not on the list.
```

Also, missing link?

[origin]/api/auth/callback/[provider]
```

In this case, given we want to try our authentication working locally on our machine and we're using **GitHub** as our OAuth provider, it'll be:

<Tabs groupId="frameworks">
  <TabItem value="next" label="Next.js" default>

```
http://localhost:3000/api/auth/callback/github
```

:::info

> NextAuth.js will already create this API endpoint for you when we start the application later. Note that because we're using Next.js, locally it starts our server on port `3000` by default. Hence, the origin is `http://localhost:3000`.

Reviewer 2: ```suggestion
Auth.js will already create this API endpoint for you when we start the application later. Note that because we're using Next.js, locally it starts our server on port `3000` by default. Hence, the origin is `http://localhost:3000`.
```

</Tabs>
:::info
The last part of the URL, `[provider]`, is the ID of the provider you're using. In this case, we're using GitHub, so it's `github`. If you're using Google, it'll be `google`, etc... We keep track of the provider IDs internally.
The same id is used in the `signIn()` method we saw earlier.
:::
To register, tap on "Register application" button.

The next screen shows all the configurations for your newly created OAuth app. For now, we need two things from it - the **Client ID** and **Client Secret**:

<img src={gettingClientIdSecretImg} />

The Client ID is always there, a public identifier of your OAuth application within GitHub. Click on the **Generate a new client Secret** button and should be presented with a new string (which is just a randomized string).

:::warning
Keep both your Client ID and Client Secret secure and never expose them to the public or share them with people outside your organization. With them, a malicious actor could hijack your application and cause you and your user serious problems!
:::

Cool! We have finished configuring our OAuth provider, now let's wire all together so we can finally see authentication working in our app!

:::info

> As noted previously, NextAuth.js has built-in support for multiple OAuth providers, <a href="">here is the full list</a>. You can also easily build your own in case the provider you need is not on the list.

Reviewer 2: ```suggestion
As noted previously, Auth.js has built-in support for multiple OAuth providers, <a href="">here is the full list</a>. You can also easily build your own if the provider you need is not on the list.
```

$ npm run next dev
```

  </TabItem>
  <TabItem value="sveltekit" label="SvelteKit">
TODO SvelteKit
  </TabItem>
  <TabItem value="solidstart" label="SolidStart">
TODO SolidStart
  </TabItem>
</Tabs>

You should see the following page:

<img src={startAppAndSignInImg} />

Click on "Sign in" and then on "Sign in with GitHub": Auth.js will redirect you to GitHub, and GitHub will recognize our app [that we just registered](#2-configuring-oauth-provider) and ask the user (in this case you) to enter its credentials to proceed:

<img src={githubAuthCredentials} />

> Once inserted and correct, GitHub will redirect the user to our app and Auth.js will take care of any further calls with GitHub to get access to the user profile and start a user session safely in the background:

Reviewer: ```suggestion
Once inserted and correct, GitHub will redirect the user to our app and NextAuth.js will take care of any further calls with GitHub to get access to the user profile and start a user session safely in the background:
```

<img src={nextAuthUserLoggedIn} />

Great! We have completed the whole E2E authentication flow setup so that users can log in to our application through GitHub!

:::info
You can create your Sign In page instead of using the default one from Auth.js. You can learn how to do so in our dedicated guide for it.

> :::

Reviewer: I think this is unnecessary to mention here.

$ npm run next dev
```

  </TabItem>
  <TabItem value="sveltekit" label="SvelteKit">
TODO SvelteKit
  </TabItem>
  <TabItem value="solidstart" label="SolidStart">
TODO SolidStart
  </TabItem>
  <TabItem value="core" label="Vanilla (No Framework)">
TODO Core
  </TabItem>
</Tabs>

You should see the following page:

<img src={startAppAndSignInImg} />

Click on "Sign in" and then on "Sign in with GitHub": Auth.js will redirect you to GitHub, and GitHub will recognize our app [that we just registered](#2-configuring-oauth-provider) and ask the user (in this case you) to enter its credentials to proceed:

<img src={githubAuthCredentials} />

> Once inserted and correct, GitHub will redirect the user to our app and NextAuth.js will take care of any further calls with GitHub to get access to the user profile and start a user session safely in the background:

Reviewer 2: ```suggestion
Once inserted and correct, GitHub will redirect the user to our app and Auth.js will take care of any further calls with GitHub to get access to the user profile and start a user session safely in the background:
```
